src/table.py

```diff
@@ -1959,29 +1997,37 @@ def clean_graphics():
         new_rects = []  # the final list of joined rectangles
         # -------------------------------------------------------------------------
         # Strategy: Join rectangles that "almost touch" each other,
-        # Extend first rectangle with any remaining in the list that touches it.
+        # Extend first rectangle with any other that is a "neighbor".
         # Then move it to final list and continue with the rest.
         # -------------------------------------------------------------------------
         while prects:  # the algorithm will empty this list
-            r = prects[0]  # first rectangle
+            r = +prects[0]  # first rectangle
             repeat = True
             while repeat:  # this loop extends first rect in list
                 repeat = False  # will be set to true if any other rect touches
-                for i in range(len(prects) - 1, -1, -1):  # run backwards
-                    if i == 0:  # don't touch first rectangle
-                        continue
-                    if are_neighbors(r, prects[i]):  # touches rect 0!
+                for i in range(len(prects) - 1, 0, -1):  # run backwards
+                    if are_neighbors(r, prects[i]):  # close enough to rect 0!
                         r |= prects[i]  # extend first rect
-                        prects[0] = +r  # update it in list
                         del prects[i]  # delete this rect
                         repeat = True  # check remaining
 
             # move first item over to result list
+            prects[0] = +r  # update rect 0
             new_rects.append(prects.pop(0))
             prects = sorted(list(set(prects)), key=lambda r: (r.y1, r.x0))
 
         new_rects = sorted(list(set(new_rects)), key=lambda r: (r.y1, r.x0))
-        return [r for r in new_rects if r.width > 5 and r.height > 5], paths
+        # return only rectangles that do contain some text
+        return [
+            r
+            for r in new_rects
+            if not white_spaces.issuperset(
+                page.get_textbox(
+                    r,
+                    textpage=TEXTPAGE,
+                )
+            )
+        ], paths
```

Comment:
I had to struggle quite hard before i understood what this list comprehension does. Prefer a loop for clarity. But feel free to ignore as this isn't important.

Reply:
This is the fastest version of the logic, because the comprehension is internally compiled - as opposed to an explicit loop.
Especially since the latest Python version, another performance improvement has been implemented by no longer internally creating a function object representing the comprehension.

What I might want to do is instead something entirely different - which will indeed need to be implemented via a separate loop:

What I am doing here is identifying bboxes that wrap separate vector graphics belonging together. The borders of these wrappers are later used like virtual vector graphics themselves thus helping to identify tables and cells.

If in the wrapping bbox is no (real) text, then not only the bbox, but all other vector graphics inside it should be likewise ignored - i.e. deleted from the `paths` list. This should speed up table recognition further down.

Might be worthwhile to do this now - as long as there is no urgency b/o a new release _ante portas._

Reply:
Ok it was easier than I thought: the complex list comprehension is gone with the next update.